Firestore/Source/API/FIRDocumentSnapshot.mm

```diff
@@ -187,48 +199,97 @@ - (FieldValueOptions)optionsForServerTimestampBehavior:
-            refDatabase.database_id().c_str(), database.project_id().c_str(),
-            database.database_id().c_str());
-    }
-    const DocumentKey &key = ref.key();
-    return [[FIRDocumentReference alloc] initWithKey:key firestore:_snapshot.firestore()];
+- (id)convertedValue:(FieldValue)value options:(const FieldValueOptions &)options {
+  switch (value.type()) {
+    case FieldValue::Type::Null:
+      return [NSNull null];
+    case FieldValue::Type::Boolean:
+      return value.boolean_value() ? @YES : @NO;
+    case FieldValue::Type::Integer:
+      return @(value.integer_value());
+    case FieldValue::Type::Double:
+      return @(value.double_value());
+    case FieldValue::Type::Timestamp:
+      return [self convertedTimestamp:value options:options];
+    case FieldValue::Type::ServerTimestamp:
+      return [self convertedServerTimetamp:value options:options];
+    case FieldValue::Type::String:
+      return util::WrapNSString(value.string_value());
+    case FieldValue::Type::Blob:
+      return MakeNSData(value.blob_value());
+    case FieldValue::Type::Reference:
+      return [self convertedReference:value];
+    case FieldValue::Type::GeoPoint:
+      return MakeFIRGeoPoint(value.geo_point_value());
+    case FieldValue::Type::Array:
+      return [self convertedArray:value.array_value() options:options];
+    case FieldValue::Type::Object:
+      return [self convertedObject:value.object_value() options:options];
+  }
+
+  UNREACHABLE();
+}
+
+- (id)convertedTimestamp:(const FieldValue &)value options:(const FieldValueOptions &)options {
+  FIRTimestamp *wrapped = MakeFIRTimestamp(value.timestamp_value());
+  if (options.timestamps_in_snapshots_enabled()) {
+    return wrapped;
   } else {
-    return [value valueWithOptions:options];
+    return [wrapped dateValue];
   }
 }
 
-- (NSDictionary<NSString *, id> *)convertedObject:(FSTObjectValue *)objectValue
-                                          options:(const FieldValueOptions &)options {
-  NSMutableDictionary *result = [NSMutableDictionary dictionary];
-  [objectValue.internalValue
-      enumerateKeysAndObjectsUsingBlock:^(NSString *key, FSTFieldValue *value, BOOL *stop) {
-        result[key] = [self convertedValue:value options:options];
-      }];
-  return result;
+- (id)convertedServerTimetamp:(const FieldValue &)value options:(const FieldValueOptions &)options {
+  const auto &sts = value.server_timestamp_value();
+  switch (options.server_timestamp_behavior()) {
+    case ServerTimestampBehavior::kNone:
+      return [NSNull null];
+    case ServerTimestampBehavior::kEstimate: {
+      FieldValue local_write_time = FieldValue::FromTimestamp(sts.local_write_time());
+      return [self convertedTimestamp:local_write_time options:options];
+    }
+    case ServerTimestampBehavior::kPrevious:
+      return sts.previous_value() ? [self convertedValue:*sts.previous_value() options:options]
+                                  : [NSNull null];
+    default:
```

Comment:
If options.server_timestamp_behavior returns an enum, then you could eliminate the default case, (and add an UNREACHABLE after the switch.) Our (c++) compiler settings will catch it for you at compile time if a default-less switch becomes non-exhaustive. I don't know if the same is true for objc... though even if it isn't, it would still be equivalent to what you've got now.